app/controllers/qsets_controller.rb

```diff
@@ -10,6 +10,36 @@ def index
     end
   end
 
+  def qset_json
```

Comment:
General practice is to overload the same `#show` route to provide html when requested and JSON when requested, so there doesn't need to be a separate endpoint for every resource that we want JSON for. That is, such that you can call `/qsets/x` requesting JSON be returned, without needing to know the specific `/qsets/x/qset_json` endpoint. To implement it, we use a few tricks to make things much smoother.

For the sake of expediency, I coded up something that works for me locally, hopefully it'll be plug-and-play.

In **qsets_controller.rb**:
```
  def show
    respond_to do |format|
      format.html { show_qset_html }
      format.json { show_qset_json }
    end
  end

  # returns ordered list of questions that have been pre-filtered based on current_user's permissions
  def get_ordered_questions
    # sorts by default by net votes; secondary sort by create date
    @questions = Question.includes(:answers).where(qset_id: @qset.id).plusminus_tally.order(created_at: :desc)
    @questions = @questions.where(user_id: current_user) if cannot? :see_all_questions, @qset
  end

  # handles the request to show different views depending on qset type
  def show_qset_html
    # check if view needs to render question counts for subsets
    if @qset.settings(:permissions).qset_type != 'questions'
      @qsets = @qset.children
      # a hash of qset question counts keyed by qset id
      @subset_question_counts = @qsets.map do |s|
        count = 0
        s.self_and_descendants.each do |q|
          scope = q.questions
          scope = scope.where(user_id: current_user) if cannot? :see_all_questions, q
          count += scope.count
        end
        [s.id, count]
      end.to_h
    end
    # check if view needs to render questions
    if @qset.settings(:permissions).qset_type != 'subsets'
      @feedback_active = !!current_user
      get_ordered_questions
      if can? :see_all_questions, @qset
        @filter_mine = true if cookies[:all_mine_other_filter] == 'mine'
        @filter_other = true if cookies[:all_mine_other_filter] == 'other'
        @filter_all = true unless @filter_mine or @filter_other
      else
        # show only the current user's questions
        cookies[:all_mine_other_filter] = 'mine'
        @filter_mine = true
      end
      # display questions view
      if @qset.settings(:permissions).qset_type == 'questions'
        render :show_questions
      end
    else
      render :show_subsets
    end
  end

  def show_qset_json
    filter = params.permit(:filter).fetch(:filter, '')
    get_ordered_questions
    if filter == 'other'
      @questions = @questions.where.not(user_id: current_user)
    elsif filter =='mine'
      @questions = @questions.where(user_id: current_user)
    end
    render json: @questions
  end
```

You need to add this new method to the **question.rb** model:  _(so that the `render json: @questions` line in the controller works as expected)_
```
  def as_json(options={})
    super(:include => {:answers => {}})
  end
```

This means you don't have to call a special endpoint anymore, so you can change this line in **qsets.js**:
```
// $.getJSON("/qsets/" + qsetid + "/qset_json?", {filter: Cookies.get('all_mine_other_filter')}, function(data) {
$.getJSON("/qsets/" + qsetid, {filter: Cookies.get('all_mine_other_filter')}, function(data) {
```
And remove the [special route](https://github.com/penzance/askup/pull/85/files#diff-21497849d8f00507c9c8dcaf6288b136R24) in *routes.rb*.

All the lines in the JS that refer to `quizQuestions.questions` can now remove the unnecessary `.questions`, and simply be `quizQuestions`, since all of our questions are now the top level of the JSON response, not embedded in a sub-object called `questions`. So:

**answers.js**
```
// var feedbackQid = quizQuestions.questions[quizAllIndex].id;
var feedbackQid = quizQuestions[quizAllIndex].id;
...
// if (quizQuestions.questions.length > (quizAllIndex + 1)){
if (quizQuestions.length > (quizAllIndex + 1)){
```

and **qsets.js**: _(note we're now getting the answer from the embedded answer in the question object itself, not from a different list)_
```
// $modal.find('.modal-title').text(quizQuestions.questions[quizAllIndex].text);
// $modal.find('.first-answer').text(quizQuestions.answers[quizAllIndex][0].text);
$modal.find('.modal-title').text(quizQuestions[quizAllIndex].text);
$modal.find('.first-answer').text(quizQuestions[quizAllIndex].answers[0].text);
```

Whew. I think that'll just about do it. If you have questions about implementation or encounter bugs then we can tackle them Monday.